File: DynamicPrograming/java/MinChange.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class MinChange {
  private MinChange() {
  }

  public static int minCoins(int amount, List<Integer> coins) {
    Map<Integer, Integer> memo = new HashMap<>();
    int answer = dfs(amount, coins, memo);
    return answer >= Integer.MAX_VALUE / 2 ? -1 : answer;
  }

  private static int dfs(int remaining, List<Integer> coins, Map<Integer, Integer> memo) {
    if (remaining < 0) {
      return Integer.MAX_VALUE / 2;
    }
    if (remaining == 0) {
      return 0;
    }
    if (memo.containsKey(remaining)) {
      return memo.get(remaining);
    }
    int best = Integer.MAX_VALUE / 2;
    for (int coin : coins) {
      best = Math.min(best, 1 + dfs(remaining - coin, coins, memo));
    }
    memo.put(remaining, best);
    return best;
  }
}

File: DynamicPrograming/csharp/MinChange.cs
using System;
using System.Collections.Generic;

namespace DynamicProgrammingSolutions;

public static class MinChange
{
    public static int MinCoins(int amount, IReadOnlyList<int> coins)
    {
        var memo = new Dictionary<int, int>();
        var answer = Dfs(amount, coins, memo);
        return answer >= int.MaxValue / 2 ? -1 : answer;
    }

    private static int Dfs(int remaining, IReadOnlyList<int> coins, IDictionary<int, int> memo)
    {
        if (remaining < 0)
        {
            return int.MaxValue / 2;
        }

        if (remaining == 0)
        {
            return 0;
        }

        if (memo.TryGetValue(remaining, out var cached))
        {
            return cached;
        }

        var best = int.MaxValue / 2;
        foreach (var coin in coins)
        {
            best = Math.Min(best, 1 + Dfs(remaining - coin, coins, memo));
        }

        memo[remaining] = best;
        return best;
    }
}
